### ISSUE TITLE
Support for "LumaCode"?

### ISSUE THREAD

**Issue by c0pperdragon** [marqs85/ossc#108]
Earlier this year, I came up with the idea of creating a proprietary video format to get digital video data from certain retro computers for further upscaling:  https://github.com/c0pperdragon/LumaCode/wiki/Overview
User feedback is pretty good but I guess for users already owning a OSSC, the mandatory RGBtoHDMI seems a bit overkill. 
So maybe you are interested in adding support for lumacode into the OSSC, as I guess this should be possible with the existing hardware.

**Comment by marqs85** [comment]
The idea is definitely interesting and could be supported on OSSC, but robust decoding of multi-sample codes would require horizontal sampling rate and phase to be adjusted accurately to avoid wrong colors / artifacts.

**Comment by c0pperdragon** [comment]
Yes, this will most likely require individual fine tuning, at least for the Atari 8-bit machines what need the highest data rate of all the supported computers. The C64 and and machines with lower data rate normally work out of the box without any re-adjustments. Just by telling the RGBtoHDMI that the source is in fact such a lumacode signal. Of course, I have no idea how to integrate such a feature into your user interface.

**Comment by Cod3N3rd** [comment]
Hi,
Did you move forward with this?
I just bought the ossc and took for granted that lumacode was supported.. (obviously wrong).
Looks like lumacode got some traction, at least with atari 8-bit community, and it would be great to see it supported by ossc.

**Comment by marqs85** [comment]
This is not yet moved forward, unfortunately. I could create a test version for a configuration that already operates at 2x oversampling, but due to lack of suitable HW and time my debugging capabilities are currently limited.

**Comment by c0pperdragon** [comment]
@marqs85 
If this helps, I could extend the firmware on my signal generator to support all lumacode patterns and just send it to you. 
So you would not have to get all those 1980ties machines and also Lumacode-enable them.

**Comment by marqs85** [comment]
Let me first create a test firmwared for you to try. If that doesn't work out, then test hardware with pattern generation could be useful.

**Comment by marqs85** [comment]
A test firmware can be now downloaded [here](http://www.infocult.com/m/tmp/ossc_1.10-luma.bin). It is configured for C64 so the following settings should be applied:

Input: AV2_RGsB (RCA from C64 connected to green Y jack)
240p/288p proc: Line2x
Line2x mode: 384x240 optim.
H. samplerate: 504 for PAL C64, 512/520 for NTSC depending on VIC-II model
Sampling phase: adjusted manually until colors correct and stable

**Comment by c0pperdragon** [comment]
This is totally awesome, that you start to integrate lumacode support into the OSSC. I very much think there is a huge number of existing OSSC users that would love to get this option. Unluckily I do not have an OSSC of my own (went straight to the RGBotHDMI option instead).

But I know of a few prominent users with a github account that have both an OSSC and a VICIIdizer. So maybe someone here can give this a try:
@TheRetroChannel
@JanBeta

I don't know if @randyrossi , who works on the Kawari (a VIC-II replacment for the C64) with possible support for Lumacode output does have an OSSC, but his input would also be very valuable.

**Comment by TheRetroChannel** [comment]
I don't have an OSSC, only the tink 5x. But I may be able borrow one to do some testing

**Comment by marqs85** [comment]
> I don't have an OSSC, only the tink 5x. But I may be able borrow one to do some testing

That would be great. Let me know if the test firmware works as expected.

**Comment by TheRetroChannel** [comment]
@marqs85 Got my hands on an OSSC, but it looks like the above link is dead. Happy to test it out if you repost it

**Comment by marqs85** [comment]
The link works for me now, perhaps there was temporary downtime on the server.

**Comment by TheRetroChannel** [comment]
Alright, the download worked and I got the OSSC updated. It took me a little while as I didn't realise the new firmwares do not work with the older hardware versions - I think this one was 1.16. Anyway, did the hardware mod and got it all working. But the lumacode output is not working as expected. I can get it to sync and look almost right, but after messing around with the sampling phase and playing with the offset and gain settings I could never get a perfect image.

Here's an example https://youtu.be/UREFcGgG_vU

**Comment by marqs85** [comment]
Ok, so aside from sampling jitter the functionality seems to work. I'll need to see if multiplying the sample rate further could improve stability as the video ADC does not perform very well at low frequencies.

**Comment by marqs85** [comment]
I've updated the linked test firmware with some improvements, hope that helps.

**Comment by TheRetroChannel** [comment]
Thanks marqs85. The new firmware does seem to help with sampling. I can set the phase to 213deg and get a stable image, but the colours are wrong. Playing with the G/Y offset does affect the colours but they never end up correct. I don't know if that's a matter of swapping some palette settings of if that's another issue. Here's another quick video https://youtu.be/EMAyqLI3B1A

Cheers, Mark

**Comment by marqs85** [comment]
The file has been updated again, hopefully this time everything works. I also added an option under Video in proc where Lumacode now has to be explicitly enabled (C64 and Spectrum palettes are currently supported).

**Comment by TheRetroChannel** [comment]
Getting close!
![Screenshot 2024-06-01 11-38-26](https://github.com/marqs85/ossc/assets/100115197/8f51ab67-ba11-4ce0-8636-5257210ff8dd)

Just the colours are slightly off, especially with red/brown/orange. Here's an example with the same machine, OSSC on the left, s-video on the right:
![Screenshot 2024-06-01 120120](https://github.com/marqs85/ossc/assets/100115197/053dc46b-cfba-4328-a639-2f1a0f4e1dd5)

Note these were all done on a PAL C64 6569R3 (8 luma level) VIC-II with the following settings:
Video in proc: Lumacode C64
Line2x mode: 384x240 optim.
H. samplerate: 504
Sampling phase: 22deg

I'm not sure if there's a way to adjust the OSSC to output a 4:3 ratio in this mode, or if I just need to set my capture device to squeeze it back down - again I'm still learning my way around the OSSC.

I don't have Lumacode for the Spectrum so can't test that. I do however have both the older 5 luma level VIC-IIs for PAL and NTSC which I can test, the only major one I don't have is the 8 luma level NTSC VIC-II.

**Comment by marqs85** [comment]
I used color codes listed [here](https://www.davenicholson.ca/2018/06/commodore-64-color-codes.html) but it seems order of orange and brown there are swapped. I fixed and uploaded the firmware again. You can try e.g. Line5x with 384x240 mode and 1920x1200/1600x1200 Line5x format to see if that results to more accurate aspect.

**Comment by TheRetroChannel** [comment]
Yep, using 5x with 1920x1200 looks about right. And orange and brown are now correct. 

I think the overall colour palette could still use some tweaking. This one here is the most accurate I've found https://jamontoads.net/p/lumachroma.html (the hex values are towards the end of the page).

**Comment by marqs85** [comment]
The palette is now updated with those values.

**Comment by TheRetroChannel** [comment]
Finally got a chance to test out the latest update. It looks soo much better - thanks Markus!

For a comparison, svideo on the left, new palette centre, old palette right.

![image](https://github.com/marqs85/ossc/assets/100115197/4357cd7a-5ec2-446a-a9e9-556981f03797)

Of course they don't totally match the s-video output but that's likely more down to my settings with the tink 5x and subtle VIC-II differences. I think the new palette is a pretty damn accurate.

I'll do a few tests with other C64/128 and VIC-II combinations and report back shortly

**Comment by TheRetroChannel** [comment]
Alright, this probably won't come as a surprise to those familar with the C64 lumacode. It works with all PAL revisions - long and shortboard, and also works with PAL 128s. I cannot test NTSC as I only have the 6567R56A which does not work correctly with lumacode without a custom firmware from @c0pperdragon at least I assume this is still the case?

As for the rest of the NTSC stuff, I'm unable to test. I guess if we stick with the current palette then it may look a bit weird to NTSC users as the NTSC palette is a bit different. But all the best C64 stuff is coded for PAL so who cares ;-P

**Comment by marqs85** [comment]
The core functionality is now integrated in e49106b6358d86cf2e298a3943666a66909e67d1. Closing the issue for now. Support for other systems (like NES) may be added once color encoding details are published.

**Comment by marqs85** [comment]
A new test version is [available](http://www.infocult.com/m/tmp/ossc_1.11-lc.bin) with preliminary support for NES added (Line2x or 5x 256col mode). It also restores the more accurate C64 palette. The code was restructured to be able to better support larger amount of systems, hopefully that didn't break Lumacode decoding for existing sources.

**Comment by TheRetroChannel** [comment]
C64 colour palette now looks correct. No other issues found there.

NES is working, the colour palette could use some tweaking, but unlike the 64 I don't have any suggestions for this. 

Some pics of SMB1 with composite vs OSSC
Composite
![Screenshot 2024-07-09 15-34-32](https://github.com/marqs85/ossc/assets/100115197/3124f6c8-6909-4301-b3ed-4f06f591b3c7)
OSSC
![Screenshot 2024-07-09 15-28-04](https://github.com/marqs85/ossc/assets/100115197/87d61fed-ec2c-4466-8b97-4199e310fafe)

Also it looks like one of the colour emphasis bits is not working. From what I understand there are 3 emphasis bits that can be set, and according to [this utility](https://forums.nesdev.org/viewtopic.php?t=13264) only the first two are working - see screenshots below.

Colour palette/emphasis tests
Composite - emphasis OFF
![Screenshot 2024-07-09 14-04-56](https://github.com/marqs85/ossc/assets/100115197/be11bbde-f261-4b62-ad43-bb250e9310fe)
OSSC - emphasis OFF
![Screenshot 2024-07-09 14-02-56](https://github.com/marqs85/ossc/assets/100115197/3c10d86d-d8c6-4168-bb90-59ef31ab16b8)
Composite - emphasis 1,0,0
![Screenshot 2024-07-09 14-05-00](https://github.com/marqs85/ossc/assets/100115197/96595801-8622-4d71-aeed-100dddf437b7)
OSSC - emphasis 1,0,0
![Screenshot 2024-07-09 14-02-59](https://github.com/marqs85/ossc/assets/100115197/9c0ed603-cca5-419e-8513-4cb270d8aa98)
Composite - emphasis 1,1,0
![Screenshot 2024-07-09 14-05-02](https://github.com/marqs85/ossc/assets/100115197/5bdc6598-becf-4725-a74b-cdce4bfcfcc3)
OSSC - emphasis 1,1,0
![Screenshot 2024-07-09 14-03-03](https://github.com/marqs85/ossc/assets/100115197/5646e9b2-ae56-427e-a32c-cc3b613d320b)
Composite - emphasis 1,1,1
![Screenshot 2024-07-09 14-05-04](https://github.com/marqs85/ossc/assets/100115197/302b34ad-b038-4dea-83db-8c69839d4363)
OSSC - emphasis 1,1,1 (looks the same as 1,1,0)
![Screenshot 2024-07-09 14-03-05](https://github.com/marqs85/ossc/assets/100115197/f80e4474-165b-4cb8-904d-3b72ea5d737e)

Everything else seemed to work fine. Again I found the sampling phase works best around 0 degrees. And bumping up the G/Y OFFSET to around 140 removed any noise.

**Comment by marqs85** [comment]
NES palette is set to match 2C02 PPU as defined on [this](https://www.nesdev.org/wiki/PPU_palettes#2C02) page. The code did not take into account cumulative effect of multiple emphasis bits which explains the lack of difference between 1,1,0 and 1,1,1. This is now fixed and the linked FW file is updated.

**Comment by c0pperdragon** [comment]
I have bin watching this thread with great interest. Nice to see support for lumacode on the OSSC. 

@TheRetroChannel
Mark, can you already confirm that the color emphasis bits are now working? In this case I plan to start selling the PAL variant. 
Making it compatible to the NTSC PPU will take some investigation due to some very strange behaviour in relation to the OAMADDR bug.

**Comment by TheRetroChannel** [comment]
Emphasis bits appear to be working with the lastest update.
![IMG_20240711_184725](https://github.com/marqs85/ossc/assets/100115197/9a5f195d-1681-490b-af7e-ab2721ac1a76)
I haven't found any other issues apart from the colour palette which I know is very subjective. Personally I like the look of the Kitrinx34 High Saturation, or the YUV palettes from [here](https://github.com/MiSTer-devel/NES_MiSTer/tree/master/Palette).

@c0pperdragon Do you know which palette is being used on the RGBtoHDMI? I think it also looked pretty nice

**Comment by c0pperdragon** [comment]
Thank you for confirming that the emphasis bits now work.

I took the palette for the RGBtoHDMI from this wikipedia page: 
https://en.wikipedia.org/wiki/List_of_video_game_console_palettes#Nintendo
But in the end, we should provide multiple palettes to chose from.

**Comment by wiesl** [comment]
Can you please also add support for TMS9918A/TMS9928A/TMS9929A versions with TMSdigitizer.

Thnx.

**Comment by marqs85** [comment]
It's already supported with MSX/Coleco setting.

**Comment by wiesl** [comment]
From a software side, yes.

The hardware version e.g. from https://www.tindie.com/products/c0pperdragon/rgbtohdmi-mono-lumacode/ is missing.

Documentation about Lumacode connection would also be nice.

**Comment by marqs85** [comment]
Documentation on the Lumacode HW is found on its own project space, see e.g. [here](https://github.com/c0pperdragon/LumaCode/wiki/TMSdigitizer-(for-ColecoVision-and-MSX)).

OSSC connection and related settings are shortly described in its [wikipage](https://junkerhq.net//xrgb/index.php?title=OSSC#Lumacode).

**Comment by rattboi** [comment]
are the C64 lumacode settings appropriate for VIC-20 lumacode as well? I have a basic signal but am having no luck dialing in anything around the 384x240 optimized values given in the wiki for C64 (No mention of VIC-20 there) 

Wanna make sure I'm not headed down the wrong path, vis a vis I should be able to get something working around the C64 advanced timing values, or not at all.

**Comment by wiesl** [comment]
Thnx. I mean the Lumacode "protocol" itself. e.g. graphics&short description about signals/timing, etc.

**Comment by marqs85** [comment]
C64 Lumacode mode should be palette-wise compatible with VIC-20, buth horizontal timings are different (h.total 455 for 6560-101, 497 for 6561-101?).

@wiesl The protocol documentation should be on its own space. Closing this issue since the support has been completed from OSSC side.

**Comment by Sowden** [comment]
> This is not yet moved forward, unfortunately. I could create a test version for a configuration that already operates at 2x oversampling, but due to lack of suitable HW and time my debugging capabilities are currently limited.

Not to open this back up.  But is this still the state for Atari 8-bit computers?

**Comment by marqs85** [comment]
If someone has the board installed on suitable Atari HW and is able to help with testing & debugging, I can create a test firmware(s) like was done with NES earlier.

**Comment by Sowden** [comment]
Cool beans.  I won't be able to get my hands on one till after next year, unfortunately.  I don't have a lot of knowledge for debugging, but I can help testing and giving reports.  I;ll come back here when I have one installed in my machine.  Thanks man.

**Comment by RamsesUTL** [comment]
Atari 8-bit would be really great and I would love to help out with some testing, but it seems I have to wait until c0pperdragon starts selling again next year.

**Comment by Sowden** [comment]
Yeah, and it seems like selling them next year is right on track.  That gives me a few months to have my OSSC modded so that I can get up to the latest firmware.  I will touch back here and tag @RamsesUTL when I do

**Comment by RamsesUTL** [comment]
@marqs85 @Sowden I got a delivery from c0pperdragon today (PPUdigitizer, VICIIdizer, GTIAdigitizer, TIAdigitizer) and can start testing things in the next few days.

**Comment by Sowden** [comment]
@marqs85 @RamsesUTL Cool beans.  I got my order in the mail and is on its way to me.  But my local moder that works on my OSSC has been dragging his feet and hasn't made the small firmware upgrade mod yet 😒 I might do the mod myself by the end of the month if this continues.  But you go ahead.  I'll jump in once mine gets in the mail and I'm able to upgrade my firmware.

**Comment by RamsesUTL** [comment]
So I successfully tested my NES with the PPUdigitizer and the OSSC. Tomorrow I'll put the GTIAdigitizer in one of my Atari 800XL.

**Comment by marqs85** [comment]
A test firmware can be now downloaded [here](http://www.infocult.com/m/tmp/ossc_1.12-test.bin). The following steps need to be done to get GTIA lumacode decoded properly:
* Set ADC PLL BW to High under sync Opt (defaulted to Medium due to some compatibility issues)
* Select some of the following combinations for mode / proc
** Line6x, 320col (recommended)
** Line3x/4x/6x, 384col
** Line2x, 256col
* set H. samplerate to 456
* adjust G/Y offset and phase until stable and correct picture

**Comment by RamsesUTL** [comment]
First test looks very promising. I have a stable picture in the Basic screen but still small glitching lines on the self test screen.
More tests tomorrow ...

**Comment by Sowden** [comment]
I'm still waiting for my board to come in the mail (and trying to figure out a mod solution).  But I found a program called Atari Control Picture, and it gives a array of color bars that you can use for testing.  Its titled "acp.xex" and you can get it off of the page posted below.  Hope that helps.

http://fujinet.diller.org/ATARI/UTILS/

**Comment by RamsesUTL** [comment]
I was about to start looking for something like this, thanks!

**Comment by marqs85** [comment]
> First test looks very promising. I have a stable picture in the Basic screen but still small glitching lines on the self test screen. More tests tomorrow ...

So you mean colors were correct but there was just some glitches caused by sampling jitter? I suppose you already tried adjusting PLL BW and phase.

**Comment by RamsesUTL** [comment]
Seems like I missed that ADC PLL BW information yesterday and still had sampling jitter then. Today everything looks great.
Atari 800XL (PAL) with SubCart on LG OLED C1, my settings:
ADC PLL BW -> High
240p/288p proc -> Line6x, Line6x mode -> 320x240 optim., HDMI HDR -> On
H. samplerate -> 456, Sampling phase -> 180 deg
G/Y offset -> 145
Adjusted the picture also a little horizontally.

![IMG_6350](https://github.com/user-attachments/assets/302a91ed-05af-43a6-bff7-ba01879bfb0f)

Turned on scanlines and the PVM shadow mask and watched a few demos.
Smooth scrolling without any glitches or jittering. Looks absolutely fantastic, thanks @marqs85 !!!

**Comment by marqs85** [comment]
Do you know anyone with VCS and Lumacode? It would be good to get that tested too so the issue could be then fully closed.

**Comment by Sowden** [comment]
I don't.  But our T. A. C. O. meeting is coming up.  Thats the Texas Atari Computer Owners meeting here in Dallas.  Maybe I could get a 2600 VCS from someone.  you would need to have @c0pperdragon board, the 2600, and an OSSC to fully test it out right?  Maybe C0pperdragon could chime in and help us.

**Comment by RamsesUTL** [comment]
> Do you know anyone with VCS and Lumacode? It would be good to get that tested too so the issue could be then fully closed.

I already have the TIAdigitizer and an Atari 2600 Jr. at home. Give me a few days to put everything together. Seems like I have to desolder the TIA first and maybe put a socket in.

**Comment by RamsesUTL** [comment]
Damn, I totally forgot about how I installed a composite video mod in my Atari 2600. That one is soldered to the bottom of the TIA. So I would need to remove that first and then also desolder the TIA. Now thinking about buying a second 2600 to avoid that ....
Edit: A friend suggested a quick and dirty solution. Mighty give that a try tomorrow ;-)

**Comment by RamsesUTL** [comment]
So I just soldered the TIAdigitizer on top of the TIA to avoid removing it first. Every connection looks good and I still have video over my composite mod. Thats the one I have installed:
https://videogameperfection.com/support-files/Atari%20Video%20Encoder%20Module%202600VECr52_rel8%20(2).pdf

Sadly I don't get any usefull video out of the TIAdigitizer/OSSC combination. With some cartridges nothing happens at all, the OSSC is not recognizing any signal. With the others the OSSC seems to jump wildly through different signal resolutions/frequencies (on the internal display) and ends up with "No sync".
I wish I had also ordered one of c0pperdragons RGB2HDMI boards to check if it works over that.

**Comment by c0pperdragon** [comment]
You should be able to see if the TIAdigitizer is working but just plugging its signal into a regular composite video input (if you still have access to a TV that has that). The picture should be stable but monochrome only and with a vertical stripe pattern.

**Comment by RamsesUTL** [comment]
Ah, thanks for the hint. I have (unstable, wandering) video over composite of the RT4K. It looks even more distorted on the small CRT I have at hand.

![Image](https://github.com/user-attachments/assets/d2423bec-45af-4bad-b9cd-7f95687672e5)

**Comment by marqs85** [comment]
You could also test the composite video into OSSC's Y-jack (unless you already did that when saying you got video). I also just updated the linked firmware image with support for VCS palette added.

**Comment by RamsesUTL** [comment]
I'll try the new firmware in a moment but I suspect some kind of problem with my installation or the combination of the TIAdigitizer with my composite mod.

**Comment by RamsesUTL** [comment]
New firmware is running but the problem remains that I sadly don't get a stable picture out the TIAdigitizer.
Any idea how to trouble shoot from this point @c0pperdragon ? Should I try to remove the composite mod?

**Comment by marqs85** [comment]
If you get it running, the following settings should be dialed for VCS:

* Line4x, 320col
* H. samplerate=456

**Comment by c0pperdragon** [comment]
As long as there is no stable picture on a regular TV, there is no hope that the OSSC can do anything with it. 
It is pretty hard to diagnose what is going on just from this picture. Maybe some inputs to the TIAdigitizer are not properly connected. Could be the CSYNC (pin 2) or the clock (pin 4). Another possibility could be some problems with the output driver, but this is quite unlikely. 
In case you have access to an oscilloscope it would be good to check those input pins. Maybe your specific mainboard provides signals that are just a bit too low to be reliably used.

**Comment by Sowden** [comment]
Hey ho guys.  So, I've finally gotten my OSSC hardware modded, and finally got my GTIAdigitizer installed in my 800XL.  I soldered the video wire to the RF pin and found ground by just soldering to the RF box itself.  I also cut off the pin to the far right that went into my RF box.  Here is what mine looks like.

![Image](https://github.com/user-attachments/assets/27d41d8b-d2b2-42da-bd67-3deca8c93e3e)

![Image](https://github.com/user-attachments/assets/53e0d882-b753-403a-9e7b-0f31e871e873)

![Image](https://github.com/user-attachments/assets/c79abb2f-3af5-4545-b7ba-50a7b5c6caca)

I'm like RamsesUTL where I've gotten off to a good start, but I'm not totally there.  I have a picture, but my two problems are that the colors aren't right and I have some vertical rolling noise going upward.  And when I do get color, I have these vertical jailbars.  Here is what mine looks like.

![Image](https://github.com/user-attachments/assets/1a6e28a6-0304-4226-9113-0d11fd2b2ffe)

![Image](https://github.com/user-attachments/assets/266564c3-4303-4357-929f-562cb373ee3f)

Sorry if it's hard to tell the colors from the glare, but the colors are looking pretty bad.  Hey @RamsesUTL, could you upload a json file for your profile for this so that I have a better starting off point?  And any other suggestions would be really great.  Thanks guys.

**Comment by marqs85** [comment]
I just updated the test firmware with some small tweaks that should enable VCS support in other LineX modes too (Line5x may not have best results, though).

@Sowden Have you also done the tweaks needed for all Lumacode configurations, i.e. setting ADC PLL BW -> High, adjustment of G/Y offset and sampling phase until picture is stable and has correct colors?

**Comment by Sowden** [comment]
@marqs85 Yes I did, I toyed around with all of the settings mentioned in this thread.  I got the background to a solid blue, but I could never get the text to be solid white.  And even with that, I have never gotten the rolling noise to go away.  And I can't seem to get a picture at all through Line6x.  I was thinking that taking off the clamps and soldering the points directly would give it a better connection, but it doesn't look like it has.  I'll be able to play with it again tomorrow.

**Comment by c0pperdragon** [comment]
Maybe the problem here is a misunderstanding: The Atari 8-bit machines (with the GTIAdigitizer) create a very different lumacode signal than the Atari 2600 (a.k.a. VCS) with the TIAdigitizer and need different decoding.

**Comment by marqs85** [comment]
Both should be supported by the FW now, although latter is untested still.

**Comment by RamsesUTL** [comment]
Hey @Sowden, have you tried my settings from above?
https://github.com/marqs85/ossc/issues/108#issuecomment-2588315346
And maybe try it with Line6x on a different device like a tv for a test?

**Comment by RamsesUTL** [comment]
> Both should be supported by the FW now, although latter is untested still.

I had no the time to figure out what I screwed up in my setup yet, but you hear back from me as soon as I have news ;-)

**Comment by Sowden** [comment]
Hey guys.  So I finally got a chance to sit down with this again.  After messing around with it, I have had massive success.  First I solved my problem with the rolling noise  by changing out the cable.  When I first tested it I was using a audio cable cause it was what I had on hand.  But I changed it out for a good composite Monster cable and it knocked that out.  But for my biggest problem, I found that I had to have my Analog sync LPF set to 2.5 MHz (max) found in the Sync opt. menu.  Once I did that, after some tinkering I got it to work!  So here are my settings on a 4K TCL 65S535 TV

Video in proc
- Video LPF 95MHz (HDTV II)
- R/Pr offset 96
- G/Y offset 141
- B/Pb offset 176
- Lumacode Atari GTIA
Sampling opt.
Adv. Timing
- H samplerate 456
- Sampleling phase 180 deg
Sync opt.
- Analog sync LPF 2.5MHz (max)
- ADC PLL BW High
Output opt.
- 240p/288p proc Line3x/4x/6x
- Line3x/4x/6x mode 320x240
- HDMI HDR flag On

![Image](https://github.com/user-attachments/assets/455d9801-c965-4236-95b2-89e63ab7b48e)

Like most everyone here says, I would also suggest using Line6x.  But you can mess around with the H and V backporch to get the framing right for all resolutions.  And the last small problem that I think I might still have is just a little bit of noise on the right side of the screen.  I've found that I can play around with the R/Pr, the G/Y, and the B/Pb options in the Video in proc menu to iron it out.  There is this great program colorx.xex that shows just for bars, and I used that to try to knock out the noise.

![Image](https://github.com/user-attachments/assets/9dcb58e2-40b8-4941-afd8-ee98a91de988)

And then my games look amazing!

![Image](https://github.com/user-attachments/assets/9a06d5d4-e3f8-46d2-97a1-8ef1283823d4)

I don't know about anyone else, but I'm happy with this beta firmware.  Are we waiting to hear about the 2600 to make it official?  Cause I'm ready for it!  Thanks a lot guys for your hard work, all around

**Comment by marqs85** [comment]
Good to hear you got it to work. Strange that R/Pr and B/Pb adjustments make a difference since those channels are ignored. I'm indeed waiting for VCS results before releasing the firmware, otherwise everything is ready.

**Comment by Sowden** [comment]
Ok, wait a minute.  I could be a total dunce here.  Am I supposed to select YPbPr or RGsB to view this?  Because I've been using YPbPr up till now.  That might be why the Pr and Pb offset might had been effecting the image.  Oh man, I think I've been doing this wrong, lol!

**Comment by marqs85** [comment]
AV2_RGsB is the correct input. Perhaps Lumacode should be made its own "input" with some settings automatically applied in a future firmware.

**Comment by RamsesUTL** [comment]
So, I finally came around to fix the problems in my Atari 2600 Jr. PAL setup. Turned out (as c0pperdragon suspected) the csync of my composite mod interfered with the signal of the TIAdigitizer. I made a switchable solution and can now decide between composite or lumacode over the former RF-socket. Works great ...

Now I'm getting a sharp stable lumacode picture over the OSSC except that the colours are wrong.
First how it should look, second lumacode:

![Image](https://github.com/user-attachments/assets/c907204b-494a-480d-b1ed-b5ed00877875)

![Image](https://github.com/user-attachments/assets/d69e511d-671b-4f7d-a541-a6b6add9b75d)

![Image](https://github.com/user-attachments/assets/3c7110f8-cff2-4978-b629-4f63481d9355)

![Image](https://github.com/user-attachments/assets/1efed919-1d8b-4354-817c-b49c094fb84c)

**Comment by c0pperdragon** [comment]
It looks like the OSSC uses the NTSC colors here. On PAL machines the colors look quite differently.

**Comment by marqs85** [comment]
For next major firmware revision there will be a few palette presets and option to load custom ones from SD. Official release won't be out for some time yet but the current palette data (also matching v1.12) can be checked [here](https://github.com/marqs85/ossc/blob/flash_imem/software/sys_controller/src/lumacode_palettes.c).

**Comment by RamsesUTL** [comment]
Sounds great. Let us know if you need some beta testing for the luma code stuff.

**Comment by c0pperdragon** [comment]
Hi @marqs85 !

In recent months I created a few more lumacode mod kits for various computers. It is no great surprise that some users also want support by the OSSC for these types of signal. I know that your development effort is focused on the OSSC Pro, but if this gets updates for lumacode, I imagine that the changes will eventually also get to the OSSC classic.

As I really expect that this was now the last system to receive lumacode, this could be a good time to wrap up the lumacode support once and for all. The only things still missing would be: 
* G7000 / Odyssey II
* Atari 7800
* All machines with a MC6847 (TRS 80 Color Computer, Dragon, Acorn Atom, Laser and some more)
* VIC 20
* C16 and Plus/4
* Sega Master System

If you want to look into this, I could provide you with a signal generator (https://github.com/c0pperdragon/LumaCode/tree/main/signalgenerator) for development purposes, as I just built another one.
All documentation of the various lumacode types is available at:  https://github.com/c0pperdragon/LumaCode/wiki

best regards,
Reinhard

**Comment by marqs85** [comment]
Hi,

Lumacode decoding logic is certainly easily portable between Pro and Classic. As long as different encodings and timings are clearly documented, it should be straightforward to add support for them.

A signal generator board would definitely help verifying the various source types. Actually I was planning to get a TIAdigitizer board as one 2600 just turned up recently, so perhaps you could bundle the signal generator on the same shipment. I'll make sure to add a note on the order about that.

**Comment by c0pperdragon** [comment]
That is great that you are considering adding support for my formats. It is absolutely not necessary to order anything, as I will just send you the signal generator and a TIAdigitizer without any charge. I would be so happy to get wider support for the lumacode system, as my current business is completely built around this.
Please let me know your shipping address with an email to shop@c0pperdragon.com
best regards,
Reinhard

**Comment by marqs85** [comment]
I've now added support for the remaining machines except C16. With Atari 7800 I was unable to get correct sampling for full horizontal width using [listed](https://github.com/c0pperdragon/LumaCode/blob/302b16f2417b18665800cfcf5902dc2e7ae45787/signalgenerator/diamond/source/SignalGenerator.vhd#L107) rate of 341, but 340.5 seemed to produce uniform result. Not sure if that's a problem in the signal generator or elsewhere.

I also added PAL palette for VCS although I can't say if the resulting colors are correct (e.g. some duplicates were present)

**Comment by c0pperdragon** [comment]
That is great to hear. I am looking forwarde to the next firmware release.

About the Atari 7800: 340.5 is actually correct. This is the result of
expanding the time for each pixel by a factor of 4/3 to stay below the
sample speed where things tend do go wrong. The signal generator does this
with some extra code to skip samples (a bit further down in the .vhdl). One
effect of this time expansion is that there is barely room for the sync
pulse now, which is dramatically shortened. TVs will have a problem with
that, but the RGBtoHMDI can handle this, and as it seems, you also got the
OSSC working with it.

Don't worry about the C16 for now. I know that I use an extremely weird way
to encode the colors (7 samples for every 2 pixels). This is also not
supported by the RGBtoHDMI and I have basically lost hope that it ever
will. So I am considering to just do the encoding in the same way as for
the MARIA with 4 samples per pixel, and having a shortened sync.
Unfortunately this would make it impossible to transmit any border color
for the left and right areas. Maybe this would not even be such an issue,
but I have to think about it and ask some users on their opinion. Or just
increase the transmission speed and hope for the best.


Am Mi., 4. März 2026 um 21:34 Uhr schrieb marqs85 ***@***.***

**Comment by marqs85** [comment]
The code is now updated in c7fa9527fac07bdd4cb6de9f9ae087e27974d087 and will be included in new firmware released soon.

**Comment by marqs85** [comment]
Closing the issue as finalized support is now available on both OSSC Classic and Pro.
